bgcval2/functions/ice.py

```diff
@@ -71,20 +71,26 @@ def calculate_ice_extent(nc, keys, **kwargs):
     areafile = get_kwarg_file(kwargs, 'areafile')
 
     maskname = kwargs.get('maskname', 'tmask')
-    if not loaded_area_and_mask:
+    if 'area' not in nc.variables.keys() and not loaded_area_and_mask:
         area, tmask, lat = load_area_and_mask(areafile, maskname)
+    else:
+        area = nc.variables['area'][:]
+        lat = nc.variables['nav_lat'][:]
+        tmask = nc.variables[keys[0]][:].squeeze().mask
 
+    print('tmask', tmask.min(), tmask.max(), tmask.sum())
     minIce = kwargs.get('minIce', 0.15)
     region = kwargs.get('region', 'Global')
     region = region.lower().replace('-', '').replace(' ', '').replace('_', '')
 
     # create mask:
-    new_mask = (tmask==0) + (nc.variables[keys[0]][:].squeeze()<minIce)
+    masked_area = np.ma.masked_where(tmask + (nc.variables[keys[0]][:].squeeze() < minIce), area)
+
     if region in ['n', 'north', 'northern', 'northhemisphere', 'northernhemisphere']:
-        new_mask += (lat<0.)
+        masked_area = np.ma.masked_where(masked_area.mask + (lat<0.), masked_area) 
```

Comment:
are you sure the overloading of `masked_where` in this manner returns the desired result? This is something I noticed in a few other places in this PR - usually, `masked_where()` works based on a condition applied to the array which you want to mask, but here `lat` has nothing to do with `masked_area`

In fact, I just tested it with:
```python
import numpy as np

lat = np.ma.array([[-2,4,-5],[33, -33, 33]], mask = [[False,True,False],[False, True, True]])
print(lat<0)

msk = np.ma.array([[-2,4,-5],[33, -33, 33]], mask = [[False,True,False],[False, True, True]])

print(np.ma.masked_where(msk.mask + (lat<0), msk))
```

which returns a Frankenmask:

```
[[True -- True]
 [False -- --]]
[[-- -- --]
 [33 -- --]]
 ```

I suggest you first build the mask from `lat<0` condition, then you add that with `|` to the other mask

Reply:
Hmmm, i have a feeling this is a deeper problem. masked_where gets used a lot in bgcval2, so it might be an whole PR just to fix this!

Reply:
let me have a look (ie a `grep`) see how this func is used elsewhere

Reply:
OK I have had a closer look and actually wrote me some pytest tests:

```python
import numpy as np
import pytest


def test_masked_where_1():
    """Test Lee interpretation of masked where."""
    times = np.array([[4., 1900.], [1960., 4.]])
    arr = np.array([[2000., 1800.], [1960., 2000.]])

    # Lee code: comparison/ensembles.py 
    times = np.ma.masked_where((times < 1850.) + (times > 1920.), times)
    # explicit: [[T, F], [F, T]] + [[F, F], [T, F]] = [[T, F], [T, T]]
    res = np.ma.array(times, mask=np.ma.array([[True, False], [True, True]], dtype=bool))
    np.testing.assert_array_equal(times, res)  # perfect!

    # arr receives exactly the same mask as times
    arr = np.ma.masked_where((times < 1850.) + (times > 1920.), arr)
    res = np.ma.array(arr, mask=np.ma.array([[True, False], [True, True]], dtype=bool))
    np.testing.assert_array_equal(arr, res)


def test_mask_where_2():
    """Test another type of implement."""
    lat = np.ma.array([[-2, 4, -5],[33, -33, 33]],
                      mask=[[False, True, False], [False, True, True]])
    msk = np.ma.array([[-2, 4, -5], [33, -33, 33]],
                      mask=[[False, True, False], [False, True, True]])
    # explicit:
    # [[F, T, F], [F, T, T]] + [[T, F, T], [F, T, F]] = [[T, T, T], [F, T, T]]
    implement = np.ma.masked_where(msk.mask + (lat < 0), msk)
    res = np.ma.array(msk,
                      mask=[[True, True, True], [False, True, True]])
    np.testing.assert_array_equal(implement, res)
```

the way you are assembling the masks **is not** wrong - it's just you gotta be 100% sure you are asking for the thing you want (see the two cases in the pytest test). It is, however, really hard to read, and not very straightforward should we want to change things. But that's OK, hopefully we'll not have to debug these :laughing:

Reply:
it's also computationally inefficient since you have already assembled a mask (like `times` above) only to recreate it when you need it again (like in `arr` above, or `msk`)

Reply:
OK phew - was not keen on going through 296 calls of `masked_where` in bgcval2 :laughing:

Reply:
> OK phew - was not keen on going through 296 calls of `masked_where` in bgcval2 😆

Dodged a bullet there then.

Reply:
was just about to say exactly that on Whatsapp :rofl: